The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5865538287/job/15902641766) [(plain)](https://github.com/rust-lang/rust/commit/5dae5344f408780c23312ac845f1db11f1ea37ae/checks/15902641766/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_f86286a9-b0b9-4826-840d-d46434decc5b
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=npo-align
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_f86286a9-b0b9-4826-840d-d46434decc5b
GITHUB_REF=refs/pull/114845/merge
GITHUB_REF_NAME=114845/merge
GITHUB_REF_PROTECTED=false
---
........................................................................................ 1936/4189
........................................................................................ 2024/4189
........................................................................................ 2112/4189
........................................................................................ 2200/4189
.............................................................F.......................F.. 2288/4189
.....................F......................F.......................F................... 2376/4189
....F.......................F...................F...................F................... 2464/4189
F...................F...................F............................................... 2552/4189
........................................................................................ 2728/4189
........................................................................................ 2816/4189
........................................................................................ 2904/4189
........................................................................................ 2992/4189
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroI128 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroI128` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroI128>(), size_of::<Option<NonZeroI128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI128;
3 + use core::num::NonZeroI128;
  |
3 + use std::num::NonZeroI128;
  |

error[E0412]: cannot find type `NonZeroI128` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:55
  |
5 | assert_eq!(size_of::<NonZeroI128>(), size_of::<Option<NonZeroI128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI128;
3 + use core::num::NonZeroI128;
  |
3 + use std::num::NonZeroI128;
  |

error[E0412]: cannot find type `NonZeroI128` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroI128>(), align_of::<Option<NonZeroI128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI128;
3 + use core::num::NonZeroI128;
  |
3 + use std::num::NonZeroI128;
  |

error[E0412]: cannot find type `NonZeroI128` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:57
  |
6 | assert_eq!(align_of::<NonZeroI128>(), align_of::<Option<NonZeroI128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI128;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroI16 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroI16` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroI16>(), size_of::<Option<NonZeroI16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI16;
3 + use core::num::NonZeroI16;
  |
3 + use std::num::NonZeroI16;
  |

error[E0412]: cannot find type `NonZeroI16` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:54
  |
5 | assert_eq!(size_of::<NonZeroI16>(), size_of::<Option<NonZeroI16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI16;
3 + use core::num::NonZeroI16;
  |
3 + use std::num::NonZeroI16;
  |

error[E0412]: cannot find type `NonZeroI16` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroI16>(), align_of::<Option<NonZeroI16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI16;
3 + use core::num::NonZeroI16;
  |
3 + use std::num::NonZeroI16;
  |

error[E0412]: cannot find type `NonZeroI16` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:56
  |
6 | assert_eq!(align_of::<NonZeroI16>(), align_of::<Option<NonZeroI16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI16;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroI32 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroI32` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroI32>(), size_of::<Option<NonZeroI32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI32;
3 + use core::num::NonZeroI32;
  |
3 + use std::num::NonZeroI32;
  |

error[E0412]: cannot find type `NonZeroI32` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:54
  |
5 | assert_eq!(size_of::<NonZeroI32>(), size_of::<Option<NonZeroI32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI32;
3 + use core::num::NonZeroI32;
  |
3 + use std::num::NonZeroI32;
  |

error[E0412]: cannot find type `NonZeroI32` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroI32>(), align_of::<Option<NonZeroI32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI32;
3 + use core::num::NonZeroI32;
  |
3 + use std::num::NonZeroI32;
  |

error[E0412]: cannot find type `NonZeroI32` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:56
  |
6 | assert_eq!(align_of::<NonZeroI32>(), align_of::<Option<NonZeroI32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI32;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroI64 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroI64` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroI64>(), size_of::<Option<NonZeroI64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI64;
3 + use core::num::NonZeroI64;
  |
3 + use std::num::NonZeroI64;
  |

error[E0412]: cannot find type `NonZeroI64` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:54
  |
5 | assert_eq!(size_of::<NonZeroI64>(), size_of::<Option<NonZeroI64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI64;
3 + use core::num::NonZeroI64;
  |
3 + use std::num::NonZeroI64;
  |

error[E0412]: cannot find type `NonZeroI64` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroI64>(), align_of::<Option<NonZeroI64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI64;
3 + use core::num::NonZeroI64;
  |
3 + use std::num::NonZeroI64;
  |

error[E0412]: cannot find type `NonZeroI64` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:56
  |
6 | assert_eq!(align_of::<NonZeroI64>(), align_of::<Option<NonZeroI64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI64;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroI8 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroI8` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroI8>(), size_of::<Option<NonZeroI8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI8;
3 + use core::num::NonZeroI8;
  |
3 + use std::num::NonZeroI8;
  |

error[E0412]: cannot find type `NonZeroI8` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:53
  |
5 | assert_eq!(size_of::<NonZeroI8>(), size_of::<Option<NonZeroI8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI8;
3 + use core::num::NonZeroI8;
  |
3 + use std::num::NonZeroI8;
  |

error[E0412]: cannot find type `NonZeroI8` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroI8>(), align_of::<Option<NonZeroI8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI8;
3 + use core::num::NonZeroI8;
  |
3 + use std::num::NonZeroI8;
  |

error[E0412]: cannot find type `NonZeroI8` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:55
  |
6 | assert_eq!(align_of::<NonZeroI8>(), align_of::<Option<NonZeroI8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroI8;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroIsize (line 198) stdout ----
error[E0412]: cannot find type `NonZeroIsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroIsize>(), size_of::<Option<NonZeroIsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroIsize;
3 + use core::num::NonZeroIsize;
  |
3 + use std::num::NonZeroIsize;
  |

error[E0412]: cannot find type `NonZeroIsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:56
  |
5 | assert_eq!(size_of::<NonZeroIsize>(), size_of::<Option<NonZeroIsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroIsize;
3 + use core::num::NonZeroIsize;
  |
3 + use std::num::NonZeroIsize;
  |

error[E0412]: cannot find type `NonZeroIsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroIsize>(), align_of::<Option<NonZeroIsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroIsize;
3 + use core::num::NonZeroIsize;
  |
3 + use std::num::NonZeroIsize;
  |

error[E0412]: cannot find type `NonZeroIsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:58
  |
6 | assert_eq!(align_of::<NonZeroIsize>(), align_of::<Option<NonZeroIsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroIsize;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroU128 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroU128` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroU128>(), size_of::<Option<NonZeroU128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU128;
3 + use core::num::NonZeroU128;
  |
3 + use std::num::NonZeroU128;
  |

error[E0412]: cannot find type `NonZeroU128` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:55
  |
5 | assert_eq!(size_of::<NonZeroU128>(), size_of::<Option<NonZeroU128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU128;
3 + use core::num::NonZeroU128;
  |
3 + use std::num::NonZeroU128;
  |

error[E0412]: cannot find type `NonZeroU128` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroU128>(), align_of::<Option<NonZeroU128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU128;
3 + use core::num::NonZeroU128;
  |
3 + use std::num::NonZeroU128;
  |

error[E0412]: cannot find type `NonZeroU128` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:57
  |
6 | assert_eq!(align_of::<NonZeroU128>(), align_of::<Option<NonZeroU128>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU128;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroU16 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroU16` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroU16>(), size_of::<Option<NonZeroU16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU16;
3 + use core::num::NonZeroU16;
  |
3 + use std::num::NonZeroU16;
  |

error[E0412]: cannot find type `NonZeroU16` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:54
  |
5 | assert_eq!(size_of::<NonZeroU16>(), size_of::<Option<NonZeroU16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU16;
3 + use core::num::NonZeroU16;
  |
3 + use std::num::NonZeroU16;
  |

error[E0412]: cannot find type `NonZeroU16` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroU16>(), align_of::<Option<NonZeroU16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU16;
3 + use core::num::NonZeroU16;
  |
3 + use std::num::NonZeroU16;
  |

error[E0412]: cannot find type `NonZeroU16` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:56
  |
6 | assert_eq!(align_of::<NonZeroU16>(), align_of::<Option<NonZeroU16>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU16;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroU32 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroU32` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroU32>(), size_of::<Option<NonZeroU32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU32;
3 + use core::num::NonZeroU32;
  |
3 + use std::num::NonZeroU32;
  |

error[E0412]: cannot find type `NonZeroU32` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:54
  |
5 | assert_eq!(size_of::<NonZeroU32>(), size_of::<Option<NonZeroU32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU32;
3 + use core::num::NonZeroU32;
  |
3 + use std::num::NonZeroU32;
  |

error[E0412]: cannot find type `NonZeroU32` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroU32>(), align_of::<Option<NonZeroU32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU32;
3 + use core::num::NonZeroU32;
  |
3 + use std::num::NonZeroU32;
  |

error[E0412]: cannot find type `NonZeroU32` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:56
  |
6 | assert_eq!(align_of::<NonZeroU32>(), align_of::<Option<NonZeroU32>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU32;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroU64 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroU64` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroU64>(), size_of::<Option<NonZeroU64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU64;
3 + use core::num::NonZeroU64;
  |
3 + use std::num::NonZeroU64;
  |

error[E0412]: cannot find type `NonZeroU64` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:54
  |
5 | assert_eq!(size_of::<NonZeroU64>(), size_of::<Option<NonZeroU64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU64;
3 + use core::num::NonZeroU64;
  |
3 + use std::num::NonZeroU64;
  |

error[E0412]: cannot find type `NonZeroU64` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroU64>(), align_of::<Option<NonZeroU64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU64;
3 + use core::num::NonZeroU64;
  |
3 + use std::num::NonZeroU64;
  |

error[E0412]: cannot find type `NonZeroU64` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:56
  |
6 | assert_eq!(align_of::<NonZeroU64>(), align_of::<Option<NonZeroU64>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU64;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroU8 (line 198) stdout ----
error[E0412]: cannot find type `NonZeroU8` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroU8>(), size_of::<Option<NonZeroU8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU8;
3 + use core::num::NonZeroU8;
  |
3 + use std::num::NonZeroU8;
  |

error[E0412]: cannot find type `NonZeroU8` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:53
  |
5 | assert_eq!(size_of::<NonZeroU8>(), size_of::<Option<NonZeroU8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU8;
3 + use core::num::NonZeroU8;
  |
3 + use std::num::NonZeroU8;
  |

error[E0412]: cannot find type `NonZeroU8` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroU8>(), align_of::<Option<NonZeroU8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU8;
3 + use core::num::NonZeroU8;
  |
3 + use std::num::NonZeroU8;
  |

error[E0412]: cannot find type `NonZeroU8` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:55
  |
6 | assert_eq!(align_of::<NonZeroU8>(), align_of::<Option<NonZeroU8>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroU8;
---
---- library/core/src/num/nonzero.rs - num::nonzero::NonZeroUsize (line 198) stdout ----
error[E0412]: cannot find type `NonZeroUsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:22
  |
5 | assert_eq!(size_of::<NonZeroUsize>(), size_of::<Option<NonZeroUsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroUsize;
3 + use core::num::NonZeroUsize;
  |
3 + use std::num::NonZeroUsize;
  |

error[E0412]: cannot find type `NonZeroUsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:200:56
  |
5 | assert_eq!(size_of::<NonZeroUsize>(), size_of::<Option<NonZeroUsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroUsize;
3 + use core::num::NonZeroUsize;
  |
3 + use std::num::NonZeroUsize;
  |

error[E0412]: cannot find type `NonZeroUsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:23
  |
6 | assert_eq!(align_of::<NonZeroUsize>(), align_of::<Option<NonZeroUsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroUsize;
3 + use core::num::NonZeroUsize;
  |
3 + use std::num::NonZeroUsize;
  |

error[E0412]: cannot find type `NonZeroUsize` in this scope
##[error] --> library/core/src/num/nonzero.rs:201:58
  |
6 | assert_eq!(align_of::<NonZeroUsize>(), align_of::<Option<NonZeroUsize>>());
  |
help: consider importing one of these items
  |
3 + use core::num::NonZeroUsize;
---
    library/core/src/num/nonzero.rs - num::nonzero::NonZeroUsize (line 198)

test result: FAILED. 4135 passed; 12 failed; 42 ignored; 0 measured; 0 filtered out; finished in 146.29s

error: doctest failed, to rerun pass `-p core --doc`
  local time: Tue Aug 15 09:26:49 UTC 2023
  network time: Tue, 15 Aug 2023 09:26:49 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/core/src/num/nonzero.rs
+++ b/library/core/src/num/nonzero.rs
@@ -41,6 +41,20 @@ macro_rules! nonzero_integers {
             /// with the exception that `0` is not a valid instance.
             #[doc = concat!("`Option<", stringify!($Ty), ">` is guaranteed to be compatible with `", stringify!($Int), "`,")]
             /// including in FFI.
+            ///
+            /// Thanks to the [null pointer optimization],
+            #[doc = concat!("`", stringify!($Ty), "` and `Option<", stringify!($Ty), ">`")]
+            /// are guaranteed to have the same size and alignment:
+            ///
+            /// ```
+            /// # use std::mem::{size_of, align_of};
+            #[doc = concat!("use std::num::", stringify!($Ty), ";")]
+            ///
+            #[doc = concat!("assert_eq!(size_of::<", stringify!($Ty), ">(), size_of::<Option<", stringify!($Ty), ">>());")]
+            #[doc = concat!("assert_eq!(align_of::<", stringify!($Ty), ">(), align_of::<Option<", stringify!($Ty), ">>());")]
+            /// ```
+            ///
+            /// [null pointer optimization]: crate::option#representation
             #[$stability]
             #[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
             #[repr(transparent)]

--- a/library/core/src/option.rs
+++ b/library/core/src/option.rs
@@ -119,7 +119,7 @@
 //! # Representation
 //!
 //! Rust guarantees to optimize the following types `T` such that
-//! [`Option<T>`] has the same size as `T`:
+//! [`Option<T>`] has the same size and alignment as `T`:
 //!
 //! * [`Box<U>`]
 //! * `&U`

--- a/library/core/src/ptr/non_null.rs
+++ b/library/core/src/ptr/non_null.rs
@@ -43,9 +43,27 @@ use crate::slice::{self, SliceIndex};
 /// it is your responsibility to ensure that `as_mut` is never called, and `as_ptr`
 /// is never used for mutation.
 ///
+/// # Representation
+///
+/// Thanks to the [null pointer optimization],
+/// `NonNull<T>` and `Option<NonNull<T>>`
+/// are guaranteed to have the same size and alignment:
+///
+/// ```
+/// # use std::mem::{size_of, align_of};
+/// use std::ptr::NonNull;
+///
+/// assert_eq!(size_of::<NonNull<i16>>(), size_of::<Option<NonNull<i16>>>());
+/// assert_eq!(align_of::<NonNull<i16>>(), align_of::<Option<NonNull<i16>>>());
+///
+/// assert_eq!(size_of::<NonNull<str>>(), size_of::<Option<NonNull<str>>>());
+/// assert_eq!(align_of::<NonNull<str>>(), align_of::<Option<NonNull<str>>>());
+/// ```
+///
 /// [covariant]: https://doc.rust-lang.org/reference/subtyping.html
 /// [`PhantomData`]: crate::marker::PhantomData
 /// [`UnsafeCell<T>`]: crate::cell::UnsafeCell
+/// [null pointer optimization]: crate::option#representation
 #[stable(feature = "nonnull", since = "1.25.0")]
 #[repr(transparent)]
 #[rustc_layout_scalar_valid_range_start(1)]

